### ISSUE TITLE
rendering series with inconsistent x axis values such as scatter plots?

### ISSUE THREAD

**Issue by o-3000** [leeoniya/uPlot#724]
uplot is great and efficient.

Is there a way perhaps you already know of or perhaps requiring a custom plugin approach the ability to add / remove series to a uplot which new series has distinct values especially where used in say scatter or bubble plots ?

**Comment by leeoniya** [comment]
uPlot has a different data format for arbitrary x/y pairs, called `mode: 2`, but there are no built-in path renderers for this mode, so you have to write them yourself. i have a demo that shows how this works for both scatter and bubble:

https://leeoniya.github.io/uPlot/demos/scatter.html

are you looking to add a scatter series to an existing `mode: 1` timeseries? you should be able to do this as well by implementing your own path builder, and hooking up the quadtree for hover, but it's more work. i'm not sure if i have a demo for this yet, :thinking:. if you don't need it to act a like a real series in the legend, you can use `draw` hook that does the rendering of the scatter points, but it will be more limited.

do you have an example image of what you're looking to do?

**Comment by o-3000** [comment]
Hi @leeoniya thanks for your prompt reply.

I found a work around in the interim which is to use default mode but restructure input data to conform to the x/y series sets in 2D arrays.  The only issue with this is that null x values can appear and so can null y values for which we'll need to clean up to suppress output. 

I have a lot more data to render (text labels on each point, tooltip data with extended info on each point, lines connecting each point with additional series (or custom draw hook method if series won't work). 

I am seeking to render simple numeric values for x instead of time.

I attached a screenshot of a sample uPlot prototype showing two homogenized series into one 2D array for use with uplot series plotting using 

![image](https://user-images.githubusercontent.com/96260248/181567393-31de8866-7be7-4a83-940d-79dfe164c3e8.png)

The opts data and the input data transformed to work with series shown below in debugging output but not sure we'll be able to use this approach for data transformation for every point or every case on graph functionality such as axis, lines, heatmap tiles without gaps in the shapes as the demos have etc. Such as text labels - unless we use a specific row for each point (column).

![image](https://user-images.githubusercontent.com/96260248/181567808-e52d8439-8563-4a31-8edb-cb422cdd1fb5.png)
